developer_support_script/check_build_stability.py

```diff
@@ -1,45 +1,162 @@
+"""
+A utility script to retrieve system tests from a jenkins job.
+
+The class currently acts as a singleton.
+"""
+
+# Dependencies
+import http.client as http
 import json
 import urllib.request
 from collections import Counter
 
-system_tests_url = "https://epics-jenkins.isis.rl.ac.uk/job/System_Tests/{}/testReport/api/json"
 
-test_metadata = "https://epics-jenkins.isis.rl.ac.uk/job/System_Tests/api/json"
+class SystemTestData:
+    """
+    Singleton class to host system test data pulled down from jenkins
+    """
+
+    def __init__(self, system_tests_url, test_metadata, test_return_quantity):
+        self.system_tests_url = system_tests_url
+        self.raw_test_metadata = test_metadata
+        self.test_return_quantity = test_return_quantity
+
+        self._json_test_metadata = self.request_page_and_metadata()
+        self._raw_builds = self.get_all_builds()
+        self._current_build, self._complete_build = self.get_current_and_complete_builds()
+        self._proccessed_builds = self.remove_current_build_if_no_results()
+        self.failing_tests = self.retrieve_failed_tests_from_builds()
+        self.top_n_failing_tests = self.get_top_n_failed_tests()
+
+    def request_page_and_metadata(self):
+        """
+        Request a jenkins webpage and return
+        the metadata json containing system test build info
+
+        :return: the metadata json
+        """
+        try:
+            page = urllib.request.urlopen(self.raw_test_metadata)
```

Comment:
I'm a little sceptical of his, because I don't think urlopen requires use of close, but pylint suggests using with for both uses of urlopen.